Properly handle trailing underscores around int-float part of exponent float

**Describe the bug**
Given either line of:

```text
a=1_e2
a=1.0_e2
```
go-toml should return an error because the int-float contains a trailing underscore.

**To Reproduce**
See PR #645

```
--- FAIL: TestTOMLTest_Invalid_Float_TrailingUsExp (0.00s)
    toml_testgen_test.go:380: Input TOML:
        # trailing underscore in integer part is not allowed
        trailing-us-exp = 1_e2
        # trailing underscore in float part is not allowed
        trailing-us-exp2 = 1.2_e2
    toml_testgen_test.go:380: JSON output from unmarshal: {"trailing-us-exp":100,"trailing-us-exp2":120}
    toml_testgen_test.go:380: test did not fail
FAIL
exit status 1
FAIL    github.com/pelletier/go-toml/v2 0.105s
```

**Expected behavior**
Return an error.

**Versions**
 - go-toml: 19751e8a519333801ea0c683d530561ed71ce02b
 - go version go1.17.1 linux/amd64

--- a/decode.go
+++ b/decode.go
@@ -405,6 +405,9 @@ func checkAndRemoveUnderscoresFloats(b []byte) ([]byte, error) {
 			if !before {
 				return nil, newDecodeError(b[i-1:i+1], "number must have at least one digit between underscores")
 			}
+			if i < len(b)-1 && (b[i+1] == 'e' || b[i+1] == 'E') {
+				return nil, newDecodeError(b[i+1:i+2], "cannot have underscore before exponent")
+			}
 			before = false
 		case 'e', 'E':
 			if i < len(b)-1 && b[i+1] == '_' {

--- a/unmarshaler_test.go
+++ b/unmarshaler_test.go
@@ -204,6 +204,16 @@ func TestUnmarshal_Floats(t *testing.T) {
 				assert.True(t, math.IsNaN(v))
 			},
 		},
+		{
+			desc:  "underscore after integer part",
+			input: `1_e2`,
+			err:   true,
+		},
+		{
+			desc:  "underscore after integer part",
+			input: `1.0_e2`,
+			err:   true,
+		},
 	}
 
 	type doc struct {
@@ -215,11 +225,15 @@ func TestUnmarshal_Floats(t *testing.T) {
 		t.Run(e.desc, func(t *testing.T) {
 			doc := doc{}
 			err := toml.Unmarshal([]byte(`A = `+e.input), &doc)
-			require.NoError(t, err)
-			if e.testFn != nil {
-				e.testFn(t, doc.A)
+			if e.err {
+				require.Error(t, err)
 			} else {
-				assert.Equal(t, e.expected, doc.A)
+				require.NoError(t, err)
+				if e.testFn != nil {
+					e.testFn(t, doc.A)
+				} else {
+					assert.Equal(t, e.expected, doc.A)
+				}
 			}
 		})
 	}

